doc: document fetchurl more completely, follow doc conventions (#300429)
NixOS/nixpkgs: doc/build-helpers/fetchers.chapter.md

[]{#fetchurl}
## `fetchurl` {#sec-pkgs-fetchers-fetchurl}

`fetchurl` returns a derivation which downloads content from a given URL and stores the unaltered contents within the Nix store.

It uses {manpage}`curl(1)` internally, and allows its behaviour to be modified by specifying a few attributes in the argument to `fetchurl` (see the documentation for attributes `curlOpts`, `curlOptsList`, and `netrcPhase`).

The resulting [store path](https://nixos.org/manual/nix/stable/store/store-path) is determined by the hash given to `fetchurl`, and also the `name` (or `pname` and `version`) values.

If neither `name` nor `pname` and `version` are specified when calling `fetchurl`, it will default to using the [basename](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-baseNameOf) of `url` or the first element of `urls`.
If `pname` and `version` are specified, `fetchurl` will use those values and will ignore `name`, even if it is also specified.

### Inputs {#sec-pkgs-fetchers-fetchurl-inputs}

> `fetchurl` expects a single argument with the following attributes:

Author: It's a copy-paste from one of my first PRs where I updated the content in the manual :)
There are a bunch of other places in the manual to update the wording, but I'll do that in a follow-up PR to keep this one focused.

> `fetchurl` expects a single argument with the following attributes:

Author: I'm keeping the "attributes" bit to avoid confusion (e.g. someone thinking 'are attribute set "entries" different from their attributes?').

```suggestion
`fetchurl` requires an attribute set with the following attributes:
```

`url` (String; _optional_)
: The URL to download from.

  :::{.note}
  Either `url` or `urls` must be specified, but not both.

>   :::

Reviewer: Another note:
```suggestion
  :::
  
`fetchurl` uses a `curl` backend the all URLs of the format [specified here](https://curl.se/docs/url-syntax.html#rfc-3986-plus) are supported.
```

In this example, we'll adapt [](#ex-fetchers-fetchurl-nixpkgs-version) to append the result of running the `hello` package to the contents we download, purely to illustrate how to manipulate the content.

```nix
{ fetchurl, hello, lib }:
fetchurl {
  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.11/.version";

  nativeBuildInputs = [ hello ];

  downloadToTemp = true;
  postFetch = ''
    ${lib.getExe hello} >> $downloadedFile
    mv $downloadedFile $out
  '';

  hash = "sha256-ceooQQYmDx5+0nfg40uU3NNI2yKrixP7HZ/xLZUNv+w=";
}
```

After building the package, the resulting file will have "Hello, world!" appended to it:

```shell
$ nix-build
(output removed for clarity)
/nix/store/ifi6pp7q0ag5h7c5v9h1c1c7bhd10c7f-version

$ cat /nix/store/ifi6pp7q0ag5h7c5v9h1c1c7bhd10c7f-version
23.11
Hello, world!
```

Note that the `hash` specified in the package is different than the hash specified in [](#ex-fetchers-fetchurl-nixpkgs-version), because the contents of the output have changed (even though the actual file that was downloaded is the same).
See [](#chap-pkgs-fetchers-caveats) for more details on how to work with the `hash` attribute when the output changes.
:::

> ## `fetchzip` {#sec-pkgs-fetchers-fetchzip}

Reviewer: `fetchzip` is rather bad structured and not explaining if it has the same interface as `fetchurl.

Would be nice to have the same structure and references to fetchurl if there are overlaps / shared interfaces. This would allow `fetchzip` to be viewed as a granular page. (i.e. if we manage to split pages anytime soon)

Author: See PR description :)

> Another PR will be created to properly document fetchzip in the near future.

I have a WIP fetchzip branch, I'm waiting for this one to be merged to avoid keeping too many PRs that depend on each other open.
